From 81e4148a398b2a46b775673abc0b0082f0f9b125 Mon Sep 17 00:00:00 2001
From: Rahul Tanwar <rtanwar@maxlinear.com>
Date: Fri, 25 Jun 2021 20:00:32 +0800
Subject: [PATCH] DRVLIB_SW-4163:  Add changes to reset the watchdog on first
 watchdog trigger.

This means that first watchdog trigger will just be a warning and
second trigger actually results in reboot. First warning trigger
prints system call trace to identify what is blocking the CPU.
---
 drivers/watchdog/grx500_wdt.c | 26 +++++++++++++++++++++-----
 1 file changed, 21 insertions(+), 5 deletions(-)

diff --git a/drivers/watchdog/grx500_wdt.c b/drivers/watchdog/grx500_wdt.c
index 2e4cc6150530..b7b22c1126db 100644
--- a/drivers/watchdog/grx500_wdt.c
+++ b/drivers/watchdog/grx500_wdt.c
@@ -30,6 +30,7 @@
 #include "watchdog_core.h"
 #include <linux/mfd/syscon.h>
 #include <linux/regmap.h>
+#include <linux/nmi.h>
 
 /* WDT MACROs */
 #define WD_RESET	(1 << 7)
@@ -114,6 +115,7 @@ struct grx500_wdt_struct {
 	 */
 	int reset_count;
 	struct reset_control **resets;
+	unsigned char warning_flag;
 };
 
 #define to_grx500_wdt(wdd)	container_of(wdd, struct grx500_wdt_struct, wdd)
@@ -138,6 +140,7 @@ static int grx500wdt_start(struct watchdog_device *wdt_dev)
 	struct grx500_wdt_struct *priv;
 
 	priv = watchdog_get_drvdata(wdt_dev);
+	priv->warning_flag = 0;
 	pr_debug("[%s]:[%d] wdd=0x%p id=%d cpu=%d priv->cpu=%d\n", __func__,
 		 __LINE__, wdt_dev, wdt_dev->id, smp_processor_id(), priv->cpu);
 	if (priv->cpu == smp_processor_id()) {
@@ -327,12 +330,24 @@ static irqreturn_t grx500wdt_irq(int irqno, void *param)
 
 	/* enable this for dump data */
 /*	grx500wdt_start(grx500_wdt);*/
-	WARN_ONCE(1, " IRQ %d triggered as WDT%d Timer Overflow on CPU %d!\n",
-		irqno, grx500_wdt->id, smp_processor_id());
-
-	if (flag == 0) {
-		show_state();
+	if (priv->warning_flag == 0) {
+		grx500wdt_stop(grx500_wdt);
+		grx500wdt_start(grx500_wdt);
+		priv->warning_flag = 1;
 		flag = 1;
+		dump_stack();
+		show_regs(get_irq_regs());
+		printk("## IRQ %d perform all-CPU dump!!!\n", irqno);
+		trigger_allbutself_cpu_backtrace();
+	} else {
+		if (flag) {
+			WARN(1, "IRQ %d triggered as WDT%d Timer Overflow on CPU %d!\n",
+			     irqno, grx500_wdt->id, smp_processor_id());
+			dump_stack();
+			show_regs(get_irq_regs());
+			trigger_allbutself_cpu_backtrace();
+			flag=0;
+		}
 	}
 	return IRQ_HANDLED;
 }
@@ -447,6 +462,7 @@ static int grx500wdt_probe(struct platform_device *pdev)
 		memcpy(priv_data, priv, sizeof(*priv));
 
 		priv_data->cpu = cpu;
+		priv_data->warning_flag = 0;
 		watchdog_set_drvdata(grx500_wdt, priv_data);
 	}
 
